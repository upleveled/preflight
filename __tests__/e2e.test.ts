import { execa } from 'execa';
import pMap from 'p-map';
import { beforeAll, expect, test } from 'vitest';

const fixturesTempDir = '__tests__/fixtures/__temp';

beforeAll(
  async () => {
    await pMap(
      [
        {
          repoPath: 'upleveled/preflight-test-project-react-passing',
          dirName: 'react-passing',
          async setup(this: { dirName: string }) {
            await execa({
              cwd: `${fixturesTempDir}/${this.dirName}`,
            })`pnpm install --frozen-lockfile`;
          },
        },
        {
          repoPath: 'upleveled/preflight-test-project-next-js-passing',
          dirName: 'next-js-passing',
          async setup(this: { dirName: string }) {
            await execa({
              cwd: `${fixturesTempDir}/${this.dirName}`,
            })`pnpm install --frozen-lockfile`;
            // Run project database migrations
            await execa({
              cwd: `${fixturesTempDir}/${this.dirName}`,
            })`pnpm migrate up`;
          },
        },
      ],
      async (testRepo) => {
        await execa`git clone --depth 1 --single-branch --branch=main https://github.com/${testRepo.repoPath}.git ${fixturesTempDir}/${testRepo.dirName} --config core.autocrlf=input`;
        return testRepo.setup();
      },
      {
        // Run up to 4 concurrent test repo setups at a time
        concurrency: 4,
      },
    );
  },
  // 10 minute timeout for pnpm installation inside test repos
  600000,
);

function sortStdoutAndStripVersionNumber(stdout: string) {
  return stdout
    .replace(/(UpLeveled Preflight) v\d+\.\d+\.\d+(-\d+)?/, '$1')
    .split('\n')
    .sort((a: string, b: string) => {
      if (b.includes('UpLeveled Preflight')) return 1;
      return a < b ? -1 : 1;
    })
    .join('\n')
    .trim();
}

test('Passes in the react-passing test project', async () => {
  const { stdout, stderr } = await execa({
    cwd: `${fixturesTempDir}/react-passing`,
  })`${process.cwd()}/bin/preflight.js`;

  expect(sortStdoutAndStripVersionNumber(stdout)).toMatchSnapshot();
  expect(stderr.replace(/^\(node:\d+\) /, '')).toMatchSnapshot();
}, 30000);

test('Passes in the next-js-passing test project', async () => {
  const { stdout, stderr } = await execa({
    cwd: `${fixturesTempDir}/next-js-passing`,
  })`${process.cwd()}/bin/preflight.js`;

  expect(sortStdoutAndStripVersionNumber(stdout)).toMatchSnapshot();
  expect(stderr.replace(/^\(node:\d+\) /, '')).toMatchSnapshot();
}, 45000);
