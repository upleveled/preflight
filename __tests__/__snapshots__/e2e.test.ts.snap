// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Passes in the react-passing test project 1`] = `
"ðŸš€ UpLeveled Preflight
[33m[STARTED][39m All changes committed to Git
[33m[STARTED][39m ESLint
[33m[STARTED][39m ESLint config is latest version
[33m[STARTED][39m GitHub repo has deployed project link under About
[33m[STARTED][39m No dependencies without types
[33m[STARTED][39m No dependency problems
[33m[STARTED][39m No extraneous files committed to Git
[33m[STARTED][39m No secrets committed to Git
[33m[STARTED][39m No unused dependencies
[33m[STARTED][39m Preflight is latest version
[33m[STARTED][39m Prettier
[33m[STARTED][39m Project folder name matches correct format
[33m[STARTED][39m Stylelint
[33m[STARTED][39m Stylelint config is latest version
[33m[STARTED][39m Use single package manager
[33m[STARTED][39m node_modules/ folder ignored in Git
[[SUCCESS][39m All changes committed to Git
[[SUCCESS][39m ESLint
[[SUCCESS][39m ESLint config is latest version
[[SUCCESS][39m GitHub repo has deployed project link under About
[[SUCCESS][39m No dependencies without types
[[SUCCESS][39m No dependency problems
[[SUCCESS][39m No extraneous files committed to Git
[[SUCCESS][39m No secrets committed to Git
[[SUCCESS][39m No unused dependencies
[[SUCCESS][39m Preflight is latest version
[[SUCCESS][39m Prettier
[[SUCCESS][39m Project folder name matches correct format
[[SUCCESS][39m Stylelint
[[SUCCESS][39m Stylelint config is latest version
[[SUCCESS][39m Use single package manager
[[SUCCESS][39m node_modules/ folder ignored in Git"
`;

exports[`Passes in the react-passing test project 2`] = `""`;
